src/v/storage/log_manager.cc

```diff
@@ -429,6 +431,82 @@ ss::future<> log_manager::remove(model::ntp ntp) {
     co_await dispatch_topic_dir_deletion(topic_dir);
 }
 
+ss::future<> remove_orphan_partition_files(
+  ss::sstring topic_directory_path,
+  model::topic_namespace nt,
+  const log_manager::ntp_to_revision& topics_on_node,
+  model::revision_id& last_applied_revision) {
+    return directory_walker::walk(
+      topic_directory_path,
+      [&topics_on_node, topic_directory_path, nt, &last_applied_revision](
+        ss::directory_entry entry) -> ss::future<> {
+          auto ntp_directory_data = partition_path::parse_partition_directory(
+            entry.name);
+          if (!ntp_directory_data) {
+              return ss::now();
+          }
+          auto ntp = model::ntp(nt.ns, nt.tp, ntp_directory_data->partition_id);
+          if (ntp_directory_data->revision_id < last_applied_revision && (
+            topics_on_node.find(ntp) == topics_on_node.end()
+            || ntp_directory_data->revision_id
+                 < topics_on_node.find(ntp)->second)) {
+              auto ntp_directory = std::filesystem::path(topic_directory_path)
+                                   / std::filesystem::path(entry.name);
+              vlog(stlog.info, "Cleaning up ntp directory {} ", ntp_directory);
+              return ss::recursive_remove_directory(ntp_directory);
+          }
+          return ss::now();
+      });
+}
+
+ss::future<> log_manager::remove_orphan_files(
+  ss::sstring data_directory_path,
+  absl::flat_hash_set<model::ns> namespaces,
+  ntp_to_revision topics_on_node,
+  model::revision_id last_applied_revision) {
+    auto data_directory_exist = co_await ss::file_exists(data_directory_path);
+    if (!data_directory_exist) {
+        co_return;
+    }
+
+    for (const auto& ns : namespaces) {
```

Comment:
think this loop may affect startup time on nodes with a lot of topics * partitions. We are recursing through every ns/topic/partition on the node. wondering if this can work in the background.

Reply:
fixed